src/Compiler/Service/ServiceParseTreeWalk.fs

```diff
@@ -379,6 +379,31 @@ module SyntaxTraversal =
         and traverseSynExpr origPath (expr: SynExpr) =
             let pick = pick expr.Range
 
+            /// Sequential expressions are more likely than
+            /// most other expression kinds to be deeply nested,
+            /// e.g., in very large list or array expressions.
+            /// We treat them specially to avoid blowing the stack,
+            /// since traverseSynExpr itself is not tail-recursive.
+            let rec traverseSequentials path expr =
+                seq {
```

Comment:
I'd move the `seq` construction inside the first branch. That way the catch-all clause will be faster with no collection rewrapping.

Reply:
Yeah, I considered that, but I didn't love how the code looked. But sure, might as well.

Reply:
b4d7c0a

Reply:
That's not the solution I had it mind:P Now you're duplicating a case check and only one branch in `traverseSequentials` will ever be hit.

Reply:
```fsharp
match expr with
| SynExpr.Sequential(expr1 = expr1; expr2 = SynExpr.Sequential _ as expr2) ->
    seq {
        // It's a nested sequential expression.
        // Visit it, but make defaultTraverse do nothing,
        // since we're going to traverse its descendants ourselves.
        yield dive expr expr.Range (fun expr -> visitor.VisitExpr(path, traverseSynExpr path, (fun _ -> None), expr))
    
        // Now traverse its descendants.
        let path = SyntaxNode.SynExpr expr :: path
        yield dive expr1 expr1.Range (traverseSynExpr path)
        yield! traverseSequentials path expr2
    }
| _ ->
    // It's not a nested sequential expression.
    // Traverse it normally.
    dive expr expr.Range (traverseSynExpr path)
```

Reply:
> Now you're duplicating a case check

Hmm, yes, if you mean whether `expr2` of the outermost sequential expression is itself a sequential expression; if it is, that will be checked again in the first call to `traverseSequentials` in order to deconstruct it — but that lets us avoid ever calling `traverseSequentials` in the first place when `expr2` of the outermost sequential is not also itself a sequential expression.

> and you might still construct an unnecessary sequence.

Only if you mean when (1) the outermost expression is a sequential expression with at least one nested sequential expression (its `expr2` is a sequential expression) and (2) the _inner_ `expr2` is not also a sequential expression. In that case, there's one extra recursive call of `traverseSequentials` (and allocation of a seq object) that we could have avoided — had we looked ahead. Of course, we could look ahead one, two, three, ... $n$ levels, but there is no obvious $n$ that makes sense to me between 0 and blowing the stack that is worth saving an allocation here.

Re: the example in https://github.com/dotnet/fsharp/pull/16882#r1531174762, I don't think it would compile, since the second branch needs to have the same type as the first, which is `(range * (unit -> 'T option)) seq`:

> ```fsharp
> match expr with
> | SynExpr.Sequential(expr1 = expr1; expr2 = SynExpr.Sequential _ as expr2) ->
>     seq {
>         // It's a nested sequential expression.
>         // Visit it, but make defaultTraverse do nothing,
>         // since we're going to traverse its descendants ourselves.
>         yield dive expr expr.Range (fun expr -> visitor.VisitExpr(path, traverseSynExpr path, (fun _ -> None), expr))
>     
>         // Now traverse its descendants.
>         let path = SyntaxNode.SynExpr expr :: path
>         yield dive expr1 expr1.Range (traverseSynExpr path)
>         yield! traverseSequentials path expr2
>     }
> | _ ->
>     // It's not a nested sequential expression.
>     // Traverse it normally.
>     dive expr expr.Range (traverseSynExpr path)
> ```

(`dive` just returns a single `(range * (unit -> 'T option))`, not a seq.)

Reply:
> (dive just returns a single (range * (unit -> 'T option)), not a seq.)

Oof, my bad.

Reply:
No problem, I appreciate the extra eyes.